eventmesh-workflow-go/internal/metrics/metrics.go

```diff
@@ -0,0 +1,177 @@
+
+func Inc(name string, label string) {
+	collector := prometheusMetrics.loadCollector(name, gauge).(*prometheus.GaugeVec)
+	collector.With(prometheus.Labels{"label": label}).Inc()
+}
+
+func Add(name string, label string, val float64) {
+	collector := prometheusMetrics.loadCollector(name, gauge).(*prometheus.GaugeVec)
+	collector.With(prometheus.Labels{"label": label}).Add(val)
+}
+
+func Sub(name string, label string, val float64) {
+	collector := prometheusMetrics.loadCollector(name, gauge).(*prometheus.GaugeVec)
+	collector.With(prometheus.Labels{"label": label}).Sub(val)
+}
+
+func Dec(name string, label string) {
+	collector := prometheusMetrics.loadCollector(name, gauge).(*prometheus.GaugeVec)
+	collector.With(prometheus.Labels{"label": label}).Dec()
+}
+
+func RecordLatency(name string, label string, latency float64) {
+	collector := prometheusMetrics.loadCollector(name, histogram).(*prometheus.HistogramVec)
+	collector.With(prometheus.Labels{"label": label}).Observe(latency)
+}
+
+func getPrometheusMetricsByConfig() *Metrics {
+	config := conf.Get()
+
+	port := ""
+	if !checkMetricsConfig(config) {
+		port = defaultEndPoint
+	} else {
+		port = config.Metrics.EndpointPort
+	}
+
+	m := &Metrics{
+		counters:   make(map[string]prometheus.Collector),
+		histograms: make(map[string]prometheus.Collector),
+		port:       port,
+	}
+	m.Init()
+	return m
+}
+
+func checkMetricsConfig(config *conf.Config) bool {
+	if config == nil || len(config.Metrics.EndpointPort) == 0 {
+		return false
+	}
+	return true
+}
+
+type Metrics struct {
+	counters   map[string]prometheus.Collector
+	histograms map[string]prometheus.Collector
+	gauges     map[string]prometheus.Collector
+	lock       sync.Mutex
+	once       sync.Once
+	port       string
+}
+
+const (
+	nameSpace       = "eventmesh"
+	subSystem       = "workflow"
+	defaultEndPoint = "19090"
+)
+
+const (
+	histogram = iota
+	gauge
+)
+
+// Init try to init metrics, include exposing http endpoint
+func (p *Metrics) Init() {
+	p.once.Do(func() {
+		p.exposeEndpoint()
+	})
+}
+
+// exposeEndpoint expose http endpoint
+func (p *Metrics) exposeEndpoint() {
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
+		err := http.ListenAndServe(fmt.Sprintf(":%s", p.port), nil)
+		if err != nil {
+			log.Errorf("fail to listen prometheus endpoint port %s, err=%v", p.port, err)
+		}
+	}()
+}
+
+// loadCollector load collector by name and collectorType
+func (p *Metrics) loadCollector(name string, collectorType int) prometheus.Collector {
+	if collector := p.getCollectorByNameAndType(name, collectorType); collector != nil {
+		return collector
+	}
+	return p.registerNewCollector(name, collectorType)
+}
+
+func (p *Metrics) getCollectorByNameAndType(name string, collectorType int) prometheus.Collector {
+	switch collectorType {
+	case histogram:
+		return p.histograms[name]
+	case gauge:
+		return p.histograms[name]
+	default:
+		panic("prometheus metrics get collector error, illegal collector type")
```

Comment:
try to use error return instead of panic
@horoc

Reply:
updated.